File: C/Trees/t1.c
#include<stdio.h>
#include<stdlib.h>

struct list{
    struct list *left;
    int data;
    struct list *right;
}typedef node;

node* insert(node*, int);
void inOrder(node*);
void postOrder(node*);
void preOrder(node*);

int main(){
    node *root = NULL;
    int ch, data;
    do{
        printf("1.Insert\n2.Inorder\n3.Preorder\n4.Postorder\nEnter your choice: ");
        scanf("%d", &ch);
        switch(ch){
            case 1: printf("Enter the value you want to insert: ");
            scanf("%d", &data);
            root = insert(root, data);
            break;

            case 2: if(root==NULL){
                printf("Tree is empty\n");
            }else{
                printf("Inorder traversal of tree\n");
                inOrder(root);
            }
            break;

            case 3: if(root==NULL){
                printf("Tree is empty\n");
            }else{
                printf("Preorder traversal of tree\n");
                preOrder(root);
            }
            break;

            case 4: if(root==NULL){
                printf("Tree is empty\n");
            }else{
                printf("Postorder traversal of tree\n");
                postOrder(root);
            }
            break;
        }
    }while(ch>0 && ch<=4);
}

node* insert(node *root, int data){
    node *ptr = NULL;
    if(root==NULL){
        ptr = (node*)malloc(sizeof(node));
        ptr->left = NULL;
        ptr->data = data;
        ptr->right = NULL;
        return ptr;
    }
    else if(data < root->data){
        root->left = insert(root->left, data);
    }
    else if(data > root->data){
        root->right = insert(root->right, data);
    }
    return root;
}

void inOrder(node *root){
    if(root!=NULL){
        inOrder(root->left);
        printf("%d\t", root->data);
        inOrder(root->right);   
    }
}

void preOrder(node *root){
    if(root !=NULL){
        printf("%d\t", root->data);
        preOrder(root->left);
        preOrder(root->right);
    }
}

void postOrder(node *root){
    if(root!=NULL){
        postOrder(root->left);
        postOrder(root->right);
        printf("%d\t", root->data);
    }
}

File: Java/Tree/T1.java
/** Binary Search Tree
 * InsertNode method to insert a new node in bst
 * Inorder Traversal 
 * Preorder Traversal 
 * PostOrder Traversal
 * LevelOrder Traversal
 */
import java.util.*;
class T1{
    class Node{
        int data;
        Node left, right;

        Node(int data){
            this.data = data;
            left = right = null;
        }
    }   

    public Node insertNode(Node root, int data){
        if(root == null){
            return new Node(data);
        }
        Node child = root;
        Node parent = null;

        while(child != null){
            parent = child;
            if(data < child.data){
                child = child.left;
            }else if(data > child.data){
                child = child.right;
            }else{
                System.out.println("Data already present");
            }
        }
        if(parent.data > data){
            parent.left = new Node(data);
        }else{
            parent.right = new Node(data);
        }
        return root;
    }

    public List<Integer> inorderTraversal(Node root) {
        List<Integer> result = new ArrayList<>();
        if (root != null) {
            result.addAll(inorderTraversal(root.left));
            result.add(root.data);
            result.addAll(inorderTraversal(root.right));
        }
        return result;
    }

    public List<Integer> preorderTraversal(Node root){
        List<Integer> li = new ArrayList<>();
        if(root!=null){
            li.addAll(preorderTraversal(root.left));
            li.addAll(preorderTraversal(root.right));
            li.add(root.data);
        }
        return li;
    }
    
    public List<Integer> postorderTraversal(Node root){
        List<Integer> li = new ArrayList<>();
        if(root!=null){
            li.add(root.data);
            li.addAll(postorderTraversal(root.left));
            li.addAll(postorderTraversal(root.right));
        }
        return li;
    }

    public List<List<Integer>> levelOrder(Node root){
        Queue<Node> queue = new LinkedList<>();
        List<List<Integer>> res = new LinkedList<List<Integer>>();
        if(root == null) return res;
        queue.offer(root);

        while(!queue.isEmpty()){
            int levelNum = queue.size();
            List<Integer> li = new LinkedList<>();
            for(int i=0; i<levelNum; i++){
                if(queue.peek().left != null) queue.offer(queue.peek().left);
                if(queue.peek().right != null) queue.offer(queue.peek().right);
                li.add(queue.poll().data);
            }
            res.add(li);
        }
        return res;
    }

    public int maxDepth(Node root){
        Queue<Node> queue = new LinkedList<>();
        int level = 0;
        
        if(root == null) return 0;
        queue.add(root);

        while(!queue.isEmpty()){
            int size = queue.size();

            for(int i = 0; i < size; i++){
                Node r = queue.poll();
                if(r.left != null){
                    queue.add(r.left);
                }
                if(r.right != null){
                    queue.add(r.right);
                }
            }
            level += 1;
        }
        return level;
    }
}